### ISSUE TITLE
"semantic-line-breaks" reflow breaks

### ISSUE THREAD

**Issue by ahgraber** [rvben/rumdl#767]
### rumdl version

0.2.43

### Installation method

cargo install

### Bug description

When reflowing from sentence-per-line to semantic-line-breaks, i noticed that the rendered outputs of some files was not the same as the pre-reflow version. It appears that emphasis (`_`, `*`, `**`, `~`) spans interact with in-span quotations and lead to inconsistent and erroneous reflow.

### Steps to reproduce

`.rumdl.toml`:

```toml
[MD013]
line_length = 80
reflow = true
reflow-mode = "semantic-line-breaks"
```

### Sentences are reordered

`t.md`:

```markdown
# Notes

_The reviewers called it "a solid contribution." The paper was accepted. Publication follows in March._
```

After `rumdl check --fix t.md`:

```markdown
# Notes

_The paper was accepted._
_Publication follows in March._
_The reviewers called it "a solid contribution."_
```

The first sentence now comes last.

### Text is deleted

`t.md`:

```markdown
# Notes

_"The reviewers called it "a solid contribution." The paper was accepted."_
```

After `rumdl check --fix t.md`:

```markdown
# Notes

_The paper was accepted."_
```

`The reviewers called it "a solid contribution."` is gone from the file.

### Expected behavior

I expect that there is no reordering or data loss; i expect that the rendered output would be identical pre and post reflow for all reflows.

### Actual behavior

I had an AI assistant investigate further; it found that emphasis and quotes interact poorly:

Emphasis is required, and a segment must end at a quotation mark that the splitter treats as a sentence boundary.

| Paragraph                                                                          | Result                                                        |
| ---------------------------------------------------------------------------------- | ------------------------------------------------------------- |
| `The paper was accepted without revision. Publication follows in March.`           | Correct. Split into two lines.                                |
| `"The reviewers called it "a solid contribution." The paper was accepted."`        | Correct. Split into two lines.                                |
| `_The paper was accepted without revision. Publication follows in March._`         | Correct, though the span is closed and reopened per sentence. |
| `_The paper was accepted (after one revision.) Publication follows in March._`     | Unchanged.                                                    |
| `` _The paper was accepted `after one revision.` Publication follows in March._ `` | Unchanged.                                                    |
| `_The reviewers called it "a solid contribution." The paper was accepted._`        | **Reordered.** The quoted sentence moves to the end.          |
| `_The reviewers called it 'a solid contribution.' The paper was accepted._`        | **Reordered.** Single quotes behave the same.                 |
| `_"A solid contribution." The paper was accepted without revision._`               | **Reordered.** A leading quote does not change the symptom.   |
| `_"The reviewers called it "a solid contribution." The paper was accepted."_`      | **Deleted.** Output is `_The paper was accepted."_`           |
| `**"The reviewers called it "a solid contribution." The paper was accepted."**`    | **Deleted.** Strong emphasis behaves the same as italic.      |

Reordering is the general symptom, but nesting one quotation inside another turns it into deletion in which the out-of-sequence segment is dropped instead of appended.
Parentheses and code spans at the same position leave the paragraph alone, and a paragraph with no emphasis is always handled correctly.

## Which spans and containers are affected

All four emphasis markers behave identically, as does strikethrough.
`_a_`, `*a*`, `**a**`, `__a__`, `~~a~~`: each reorders on a single-level quote, each deletes on nested quotes.
Code spans, link text, inline HTML such as `<em>`, and `==highlight==` are unaffected.

List items, nested list items, and blockquotes all lose the text:

```markdown
- _"The reviewers called it "a solid contribution." The paper was accepted."_
```

becomes `- _The paper was accepted."_`.
Headings and table cells are left alone, since reflow does not process them.

When a paragraph holds two damaged spans, the first one loses its text and the second is left intact:

```markdown
_"She said "yes." The vote carried."_ and _"He said "no." The motion failed."_
```

becomes `_The vote carried."_ and _"He said "no." The motion failed."_`.

I have one case where both spans and the text between them disappear, leaving a single fragment, but I have not worked out what distinguishes it.

### Configuration

```toml
### Global configuration options
[global]

### List of rules to enable exclusively (if provided, ONLY these rules will run)
# enable = []

### List of rules to disable (uncomment and modify as needed)
# # Not implemented
# "MD002", # replaced by MD041
# "MD006", "MD008", "MD015", "MD016", "MD017",
disable = [
  # "MD013", # Line length
  "MD014", # Dollar signs used before commands; commands show outputs
  # "MD028", # Blank line inside blockquote
  "MD033", # Inline HTML tags
  "MD036", # Emphasis used instead of a header # oversensitive
  "MD041", # Require title
  "MD052", # Reference link # seems to not follow footnote links
  "MD060", # Table spacing # Can interfere with mdformat
]

### List of file/directory patterns to include for linting (if provided, only these will be linted)
# include = [
#    "docs/*.md",
#    "src/**/*.md",
#    "README.md"
# ]

### List of file/directory patterns to exclude from linting
exclude = [
    # Common directories to exclude
    ".git",
    ".github",
    "node_modules",
    "vendor",
    "dist",
    "build",

    # Specific files or patterns
    "CHANGELOG.md",
    "LICENSE.md",
]

### Respect .gitignore files when scanning directories (default: true)
respect_gitignore = true

### Markdown flavor/dialect (uncomment to enable)
### Options: mkdocs, gfm, commonmark
# flavor = "gfm"

### Default line length for rules that check line length
line_length = 80

#-----------------------------------------------------------------------------
### Rule-specific configurations
#-----------------------------------------------------------------------------
### Heading style
[MD003]
style = "atx"  # Use '#' Heading style

### Unordered list style
[MD004]
style = "dash"  # asterisk, plus, dash, consistent

### (Unordered) list indentation
[MD007]
# indent = 2
style = "text-aligned" # Align markers with start of parent text

### Trailing spaces
[MD009]
br_spaces = 2 # Number of spaces for hard line breaks (default: 2)
strict = false  # Remove all trailing spaces, even line breaks (default: false)

### Line length / wrap
[MD013]
line_length = 80
paragraphs = false  # Do not warn about long paragraphs
headings = false  # Do not warn about long headings
code-blocks = false # Do not warn on extended code blocks
tables = false # Do not warn about wide tables
reflow = true
# Break at sentences first, then clauses, then conjunctions, then words,
# so an over-long sentence still wraps instead of being left alone.
reflow-mode = "semantic-line-breaks"
abbreviations = ["e.g.", "i.e.", "vs.", "U.S.", "ft.", "v."]

### Avoid duplicate headings
[MD024]
siblings_only = true

### Only one H1 / title per document
[MD025]
level = 1
front_matter_title = "title"

### Ordered list
[MD029]
style = "consistent"
# style = "one" # Use all 1's for ordered lists; let renderer number

### Spacing after list markers
# [MD030]
# ul_single = 2  # Spaces after bullet for single-line items (default: 1)
# ul_multi = 2   # Spaces after bullet for multi-line items (default: 1)
# ol_single = 1  # Spaces after number for single-line items (default: 1)
# ol_multi = 1   # Spaces after number for multi-line items (default: 1)

### Blank lines around lists
[MD032]
allow-lazy-continuation = false

### Consistent breaks
[MD035]
style = "---"

### Code block style
[MD046]
style = "fenced"  # Options: "consistent", "fenced", "indented"

### Italic
[MD049]
style = "underscore"  # Options: "consistent", "asterisk", "underscore"

### Bold
[MD050]
style = "asterisk"  # Options: "consistent", "asterisk", "underscore"

### Table formatting (pipes)
[MD055]
style = "leading-and-trailing"  # Options: see below

### Table formatting (spacing)
# [MD060]
# enabled = true
# style = "aligned"
# max-width = 120  # Independent of MD013, use 120 char threshold
```

### Operating system

macOS

**Comment by ahgraber** [comment]
@rvben -- I identified two other bugs during semantic-line-breaks reflows -- would you prefer I expand them below or as distinct issues?

- A newline inside a Markdown paragraph renders as a space, so every reflowed break following punctuation inserts a space that was not in the document. e.g.: `scaling—adding` becomes `scaling— adding`.
- a space between a hyphen and a following `$` is removed. e.g.,: `Pricing is the same - $3/$15` becomes `Pricing is the same -$3/$15`,

**Comment by rvben** [comment]
The reordering and deletion you originally reported are fixed in v0.2.45. A sentence boundary falling inside an emphasis span consumed the separating space, so words around it were dropped and fragments could be reattached out of order. `rumdl fmt` now keeps every word, and the regression harness runs the real `fmt` pipeline, requires whitespace-only changes, and requires a second pass to converge.

On your question: one issue per defect is easier to fix and close, so distinct issues please. Neither of these two needs filing, though.

**Your first case is confirmed and reproduced, so this issue stays open for it.** In `semantic-line-breaks` mode only, a break placed directly after an em dash with no whitespace on either side injects a rendered space:

```markdown
The cost of scaling—adding more nodes to the cluster—grows faster than the throughput it buys you in practice.
```

becomes

```markdown
The cost of scaling—adding more nodes to the cluster—
grows faster than the throughput it buys you in practice.
```

which renders as `cluster— grows`: 20 words in, 21 out. I probed nine punctuation shapes across all three reflow modes; en dash, slash, colon, parenthesis and hyphenated words are all clean, and `normalize` and `sentence-per-line` are clean throughout. #770 is a third instance of the same family (a period orphaned after a parenthetical), so the fix is the general one: never break where the source had no whitespace, enforced by requiring the reflowed paragraph to render identically to the original.

**Your second case I could not reproduce.** I tried `- $3/$15` at nine different wrap positions in all three modes and the space never disappeared. It is likely to depend on the surrounding paragraph, so could you paste the full paragraph and the MD013 settings you ran it with? If it turns out to be distinct from the em dash case, it is worth its own issue.

**Comment by ahgraber** [comment]
Thanks so much for the quick resolution. Very much appreciate this project and your efforts!

Here's the original text in context:

> ### [Introducing Claude Sonnet 4.5 \\ Anthropic](https://www.anthropic.com/news/claude-sonnet-4-5)
> 
> Anthropic launched Claude Sonnet 4.5, touting it as "the best coding model in the world," as it outperforms both OpenAI's GPT-5-Codex and Anthropic's own Claude Opus 4.1.
> Pricing remains the same - $3/$15 per million input/output tokens.
> 
> Alongside Claude Sonnet 4.5, Anthropic have released [Claude Code 2.0](https://www.npmjs.com/package/@anthropic-ai/claude-code) and updated the Claude Code SDK to the [Claude Agent SDK \\ Anthropic](https://www.anthropic.com/engineering/building-agents-with-the-claude-agent-sdk), which is the same Agent harness that powers Claude Code:

---

I also investigated whether this was an interaction effect with other markdown formatters, but it appears to be related to span-identification within rumdl (it reads `$3/$` as a math span.  It looks like 0.2.25/0.2.26 resolved the issue.

Given rumdl 0.2.43 and 0.2.44 (all rows unchanged , reflow = true, reflow-mode = "sentence-per-line", line_length = 80:

| Original | 0.2.43 / 0.2.44 result | Reflow triggered | Space eaten | 0.2.45 / 0.2.46 result | Comment |
|---|---|:---:|:---:|---|---|
| `One. Two - $a$ x.` | `One.\nTwo -$a$ x.` | yes | yes | *(unchanged)* | math span |
| ``One. Two - `abc` x.`` | ``One.\nTwo -`abc` x.`` | yes | yes | *(unchanged)* | code span |
| `One. Two - _abc_ x.` | `One.\nTwo -_abc_ x.` | yes | yes | *(unchanged)* | emphasis span |
| `One. Two - $a$ - $b$ x.` | `One.\nTwo -$a$ - $b$ x.` | yes | yes | *(unchanged)* | only the first occurrence is affected |
| `One. Two + $a$ x.` | `One.\nTwo + $a$ x.` | yes | no | `One.\nTwo + $a$ x.` | reflow splits the sentence, space survives; `+` does not trigger |
| `One. Two word $a$ x.` | `One.\nTwo word $a$ x.` | yes | no | `One.\nTwo word $a$ x.` | reflow splits the sentence, space survives; requires a bare `-` before the span |
| `Two - $a$ x.` | *(unchanged)* | no | no | *(unchanged)* | single sentence, no split to make |
| `Pricing for the newly released model remains exactly the same - $3/$15 per million input and output tokens across every region.` | *(unchanged)* | no | no | *(unchanged)* | single sentence; at `line_length = 0` there is no width wrapping either |
| `It costs less. Pricing is the same - $3/$15 today.` | `It costs less.\nPricing is the same -$3/$15 today.` | yes | yes | *(unchanged)* | two sentences; the sentence-boundary split is what triggers it |
| `It costs less.\nPricing is the same - $3/$15 today.` | `It costs less.\nPricing is the same -$3/$15 today.` | yes | yes | *(unchanged)* | same paragraph already split across two source lines |
| `It costs less. Pricing is the same - $3/$15 per million tokens.` | `It costs less.\nPricing is the same -$3/$15 per million tokens.` | yes | yes | *(unchanged)* | originally reported case; `$3/$` parses as a math span |

**Comment by rvben** [comment]
Thanks for the paragraph in context and for that table, @ahgraber. It made the second bug reproducible, and it is what got it fixed.

One correction on your reading though: 0.2.45 and 0.2.46 did not resolve it. They stopped the space being eaten, but only by declining to reflow those lines at all. On 0.2.46:

```
One. Two + $a$ x.   ->  "One.\nTwo + $a$ x."   (reflows)
One. Two - $a$ x.   ->  "One. Two - $a$ x."    (silently unchanged)
```

The content-preservation guard caught the corrupted output and dropped the whole reflow, so the `*(unchanged)*` column in your table was the sentence split never happening, not the bug being gone.

The cause was that reflow reconstructed the gap in front of an inline span by guessing from the characters the line happened to end with, instead of reading it from the source. Guessing wrong one way deleted a real space (`- $a$`); guessing wrong the other way inserted one that was never there (`mid*word*` becoming `mid *word*`). v0.2.47 takes the gap from the source, so all six of your affected rows now reflow and keep the space:

```
It costs less.
Pricing is the same - $3/$15 per million tokens.
```

Your em dash case is fixed in v0.2.47 as well. A semantic break is now only placed where the source actually had whitespace, so `cluster—grows` stays intact and the line wraps at a real space instead.

Closing as released in v0.2.47. Thanks again for the detail, both bugs were far easier to fix with it.
